### PR TITLE
Fixes #7292 - API File Contents bug

### LINKED ISSUE #7292: API Repo Get File Content Endpoint returns non-working "blob" HTML_URL
I noticed that the URL for the API endpoint for getting the file contents of a repo is not returning a proper URL for the "html_url" field. It uses the "blob" URL that Github uses, but we don't have that as a HTML endpoint. Need to be able to access the file in HTML the same way as on Github, not just for a branch, but also a commit ID.

### PR DESCRIPTION
**Pull request by richmahn**
A few of bugs fixed with the File Content API response, one from issue #7292, but also the COMPLETE functionality that GitHub has for the GET contents API endpoint. Here are bug fixes:

1. BUG: `html_url` field had `blob/<ref>/<file>` in the URL. This is how GitHub does it, but Gitea does `src/<ref type>/<ref>/<file>`, so now the code determines the ref type give and gives the proper `html_url` (as noted in #7292)
2. FIX: The content of the file in base64 encoding was not being returned as GitHub does. Added the `encoding` and `content` field to the response (see GitHub's file content response: https://developer.github.com/v3/repos/contents/#response-if-content-is-a-file)
3. FIX: Before the Get File Contents API endpoint only took a branch as a ref, when it should take a tag and commit ID as well. This is now added and `?ref=[branch|tag|commit]` will work, and the URLs given in the response will properly preserve this.
4. FIX: Before it only worked with files as the path in the URL. Now works for directories (returns an array of FileContentsResponse objects as the response), symlinks (will give the target of the symlink in the response), and submodules (will give the submodules's git URL in the response). The symlink target, the submodule git URL, and file contents & encoding are null if they do not apply to the type of entry. Also files return in a directory list of entries will not have their content and encoding (will be null) as that would take too long to return all file contents.

Again, please see how this is closer to what GitHub does: https://developer.github.com/v3/repos/contents/#response-if-content-is-a-file, https://developer.github.com/v3/repos/contents/#response-if-content-is-a-directory, https://developer.github.com/v3/repos/contents/#response-if-content-is-a-symlink and https://developer.github.com/v3/repos/contents/#response-if-content-is-a-submodule
(only difference is that I leave the `target:`, `submodule_git_url:`, `content:` and `encoding:` properties still in the response as `null` even if not used as having a different response objects for each entry type would have been a headache and not work with swagger.)

Also note, these changes to this API endpoint shouldn't affect anyone as they haven't been released yet (see my PR for the original endpoint code: https://github.com/go-gitea/gitea/pull/6314 - milestone 1.9.0)

Below in another comment I give some examples. Compare with the old response: https://try.gitea.io/api/v1/repos/richmahn/test/contents/README

### CODE REVIEW

**Comment by richmahn** [comment]
Here are some example JSON Responses with the GET URL:

No ref: http://localhost:3000/api/v1/repos/root/test/contents/README.md
```{
  "name": "README.md",
  "path": "README.md",
  "sha": "f3b60aa0df362906d8b64a16aa2eed6c3ec48906",
  "size": 28,
  "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=master",
  "html_url": "http://localhost:3000/root/test/src/branch/master/README.md",
  "git_url": 
 
 "http://localhost:3000/api/v1/repos/root/test/git/blobs/f3b60aa0df362906d8b64a16aa2eed6c3ec48906",
  "download_url": "http://localhost:3000/root/test/raw/branch/master/README.md",
  "type": "blob",
  "encoding": "base64",
  "content": "IyB0ZXN0CiMjIHRlc3QgdjIKCnRlc3QgZGVzYw==",
  "_links": {
    "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=master",
    "git_url": 
 
 
 
 "http://localhost:3000/api/v1/repos/root/test/git/blobs/f3b60aa0df362906d8b64a16aa2eed6c3ec48906",
    "html_url": "http://localhost:3000/root/test/src/branch/master/README.md"
  }
}```

With branch: http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=develop
{
  "name": "README.md",
  "path": "README.md",
  "sha": "f3b60aa0df362906d8b64a16aa2eed6c3ec48906",
  "size": 28,
  "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=develop",
  "html_url": "http://localhost:3000/root/test/src/branch/develop/README.md",
  "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/f3b60aa0df362906d8b64a16aa2eed6c3ec48906",
  "download_url": "http://localhost:3000/root/test/raw/branch/develop/README.md",
  "type": "blob",
  "encoding": "base64",
  "content": "IyB0ZXN0CiMjIHRlc3QgdjIKCnRlc3QgZGVzYw==",
  "_links": {
    "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=develop",
    "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/f3b60aa0df362906d8b64a16aa2eed6c3ec48906",
    "html_url": "http://localhost:3000/root/test/src/branch/develop/README.md"
  }
}

Short commit ID: http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=b7b2
```{
  "name": "README.md",
  "path": "README.md",
  "sha": "311fbf6da73cdd85bcc29f4f3809cd4dc2f5cb83",
  "size": 17,
  "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=b7b2",
  "html_url": "http://localhost:3000/root/test/src/commit/b7b2828f27a39a1967e7762b0d4799fe69343652/README.md",
  "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/311fbf6da73cdd85bcc29f4f3809cd4dc2f5cb83",
  "download_url": "http://localhost:3000/root/test/raw/commit/b7b2828f27a39a1967e7762b0d4799fe69343652/README.md",
  "type": "blob",
  "encoding": "base64",
  "content": "IyB0ZXN0Cgp0ZXN0IGRlc2M=",
  "_links": {
    "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=b7b2",
    "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/311fbf6da73cdd85bcc29f4f3809cd4dc2f5cb83",
    "html_url": "http://localhost:3000/root/test/src/commit/b7b2828f27a39a1967e7762b0d4799fe69343652/README.md"
  }
}```

Long Commit: http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=b7b2828f27a39a1967e7762b0d4799fe69343652
```{
  "name": "README.md",
  "path": "README.md",
  "sha": "311fbf6da73cdd85bcc29f4f3809cd4dc2f5cb83",
  "size": 17,
  "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=b7b2828f27a39a1967e7762b0d4799fe69343652",
  "html_url": "http://localhost:3000/root/test/src/commit/b7b2828f27a39a1967e7762b0d4799fe69343652/README.md",
  "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/311fbf6da73cdd85bcc29f4f3809cd4dc2f5cb83",
  "download_url": "http://localhost:3000/root/test/raw/commit/b7b2828f27a39a1967e7762b0d4799fe69343652/README.md",
  "type": "blob",
  "encoding": "base64",
  "content": "IyB0ZXN0Cgp0ZXN0IGRlc2M=",
  "_links": {
    "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=b7b2828f27a39a1967e7762b0d4799fe69343652",
    "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/311fbf6da73cdd85bcc29f4f3809cd4dc2f5cb83",
    "html_url": "http://localhost:3000/root/test/src/commit/b7b2828f27a39a1967e7762b0d4799fe69343652/README.md"
  }
}```

Tag: http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=v1
```{
  "name": "README.md",
  "path": "README.md",
  "sha": "f03dd716384d52b861c36f88521576870a8cd9fc",
  "size": 25,
  "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=v1",
  "html_url": "http://localhost:3000/root/test/src/tag/v1/README.md",
  "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/f03dd716384d52b861c36f88521576870a8cd9fc",
  "download_url": "http://localhost:3000/root/test/raw/tag/v1/README.md",
  "type": "blob",
  "encoding": "base64",
  "content": "IyB0ZXN0CiMjIHRlc3QKCnRlc3QgZGVzYw==",
  "_links": {
    "url": "http://localhost:3000/api/v1/repos/root/test/contents/README.md?ref=v1",
    "git_url": "http://localhost:3000/api/v1/repos/root/test/git/blobs/f03dd716384d52b861c36f88521576870a8cd9fc",
    "html_url": "http://localhost:3000/root/test/src/tag/v1/README.md"
  }
}```

File: modules/git/repo.go [hunk 1]
```diff
@@ -55,6 +55,54 @@ func (repo *Repository) parsePrettyFormatLogToList(logs []byte) (*list.List, err
...
+		return "tag"
+	case RepoRefTypeInvalid:
+		return ""
+	default:
+		return ""
+	}
+}
+
+// GetRefType gets the type of the ref based on the string
+func (repo *Repository) GetRefType(ref string) RepoRefType {
```
Comment (lunny) [inline modules/git/repo.go]: Could we return ObjectType on https://github.com/go-gitea/gitea/blob/master/modules/git/repo_object.go#L14 ?

Comment (richmahn) [reply modules/git/repo.go]: Looking into it. I actually realized I did not have this at all responding like GitHub, especially when it comes to being 4 different types of objects (dir, file, symlink, submodule) that could be returned, as well well as a list of entries. So this is being redone.

Comment (richmahn) [reply modules/git/repo.go]: Added ObjectBranch and now use those as a way to get the string of a ref. Also moved my GetRefType() into this file.

**Review by zeripath — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**